profiling/src/lib.rs

```diff
@@ -844,3 +898,64 @@ extern "C" fn execute_internal(
     }
     interrupt_function(execute_data);
 }
+
+unsafe extern "C" fn alloc_profiling_malloc(len: u64) -> *mut ::libc::c_void {
+    let ptr: *mut libc::c_void;
+
+    // TODO: prepare a function pointer to use so we don't need a runtime check
+    if PREV_CUSTOM_MM_ALLOC.is_none() {
+        ptr = zend::_zend_mm_alloc(zend::zend_mm_get_heap(), len);
+    } else {
+        let prev = PREV_CUSTOM_MM_ALLOC.unwrap();
+        ptr = prev(len);
+    }
+
+    // during startup, minit, rinit, ... current_execute_data is null
+    if zend::executor_globals.current_execute_data.is_null() {
+        return ptr;
+    }
+
+    trace!("malloc({}) -> {:p}", len, ptr);
+
+    REQUEST_LOCALS.with(|cell| {
+        let locals = cell.borrow();
+
+        if !locals.profiling_enabled {
+            return;
+        }
+
+        if let Some(profiler) = PROFILER.lock().unwrap().as_ref() {
```

Comment:
I don't think this lock is strictly necessary. It was done to appease Rust's memory model in a safe way, and the cost was acceptable because we didn't lock it very often with time profiling. This PR changes that.

We don't have to fix it in this specific PR, though. I'm just pointing out that for production I suspect we'll need to revisit this.